Only create checkpoints for unstable blocks

KtorZ opened the pull request:
# Issue Number

<!-- Put here a reference to the issue this PR relates to and which requirements it tackles -->

#639

# Overview

<!-- Detail in a few bullet points the work accomplished in this PR -->

- [x] I have only stored checkpoints for unstable blocks during restoration (and, each last block of each batch) cf: [6a7e9ff5](https://github.com/input-output-hk/cardano-wallet/commit/6a7e9ff580fff39a6f7e816b2c17c68ef02646ec)

- [x] I have fetch the `epochStability` (a.k.a `k` parameter) from genesis block (see  [2d0c5584](https://github.com/input-output-hk/cardano-wallet/commit/2d0c55845711704db5737992ed914a453432cd09)

- [x] I have reworked `applyBlock` so it returns a `NonEmpty` list of checkpoints (since it gets a `NonEmpty` list of blocks as input and, there's a 1:1 mapping between blocks and checkpoints here). This remove some awkwardness when accessing the last checkpoint of the list during restoration. (See [c2d193af](https://github.com/input-output-hk/cardano-wallet/commit/c2d193afaebf89fe32fbebd333151066a8aaa629)

# Comments

<!-- Additional comments or screenshots to attach if any -->

I've re-run the restoration benchmark which now behaves as expected. The last `k` blocks are longer to restore because a checkpoint needs to be created for each of them, otherwise, we only create a checkpoint every epoch. 

```
All results:                            
  restore testnet seq: 12.16 s          
  restore testnet 10% ownership: 12.14 s
Benchmark restore: FINISH               
Completed 2 action(s).                  
```

This should coincidentally fix the nightly builds.

<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

KtorZ committed:
only store checkpoints for unstable blocks during restoration

We restore blocks by batches, but storing a checkpoint for every single block is very
expensive: storing a checkpoint is roughly 10 various db queries, so it's 200k requests
per epoch, each of them likely useless since we can't rollback that far!

Instead, we only store a checkpoint for the each last block in a batch or, for unstable
blocks.

KtorZ committed:
do not hard-code epoch stability in 'Cardano.Wallet' but fetch it from network

KtorZ committed:
make 'applyBlocks' return a non empty list

This makes call it clear that calls for 'init' and 'last' are safe in this context.
We give a non empty list of blocks, so we get a non empty list of checkpoints

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Model.hs:
```diff
@@ -231,11 +233,14 @@ applyBlock !b (Wallet !u !pending _ s bh) =
+    -- NOTE:
+    -- The list of blocks is of length `n > 0`, scanl generates a list of length
+    -- `n+1`, so its tail is necessarily of length `n > 0`!
+    NE.fromList $ NE.tail $ NE.scanl (flip applyBlock . snd) (mempty, cp0) blocks
```
We can avoid the partial function (and comment) here by applying the first block separately:
```hs
applyBlocks (block :| blocks) cp0 =                                                                                                                                                   
    NE.scanl (flip applyBlock . snd) (applyBlock block cp0) blocks      
```

KtorZ replied:
Fair point :+1:

KtorZ replied:
Change done!

jonathanknowles commented on lib/core/test/unit/Cardano/WalletSpec.hs:
```diff
@@ -420,7 +420,11 @@ setupFixture (wid, wname, wstate) = do
     block0Date = StartTime $ posixSecondsToUTCTime 0
 
+    k :: Quantity "block" Word32
+    k = Quantity 2160
```
This stability constant (`2160`) appears to be defined in a few places. Could it be defined in a single place and imported?

Also, would it make the code more readable if we call it `epochStability` or something similar? (I did happen to know what `k` is, but perhaps others reading the code might not?)

KtorZ replied:
> Could it be defined in a single place and imported?

Yes and no. It would be wrong to hard-code it in one place (although it is in `http-bridge@.../Compatibility.hs`. Here in this test scenario (like in all scenarios where it's hard-coded), its actual value doesn't matter. Could be 42, could be 87541. Force of the habit and lack of originality made us choose 2160. Perhaps an `Arbitrary` would better convey the idea.

KtorZ replied:
*I've changed the declaration of `BlockchainParameters` to use named record fields which removes the need for intermediate variable and make things more readable:

```hs
    bp = BlockchainParameters
        { getGenesisBlock = block0
        , getGenesisBlockDate = StartTime $ posixSecondsToUTCTime 0
        , getFeePolicy = LinearFee (Quantity 14) (Quantity 42)
        , getSlotLength = SlotLength 1
        , getEpochLength = EpochLength 21600
        , getTxMaxSize = Quantity 8192
        , getEpochStability = Quantity 2160
        }
```

jonathanknowles replied:
> It would be wrong to hard-code it in one place

Why? If we intend it to be the same value in each of these places, then why is it wrong to just write it once and then re-import it each time?

Of course, if it's an arbitrary value, then why not just generate arbitrary values when needed?

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -692,34 +696,40 @@ newWalletLayer tracer bp db nw tl = do
+                    let meta' = calculateMetadata (view #slotId $ currentTip cp)
+                    DB.putCheckpoint db (PrimaryKey wid) cp
+                    DB.putTxHistory db (PrimaryKey wid) txs
+                    DB.putWalletMeta db (PrimaryKey wid) meta'
```
Is this really correct?

While it is the case that we only need to write checkpoints for **unstable** blocks, we still need to write the TX history and metadata for **all** blocks, right?

KtorZ replied:
Yes but the mistake is actually a bit below line 717:

```hs
             putCheckpoint (snd cpLast) (fst cpLast) 
```

This should be:

```hs
             putCheckpoint (snd cpLast) newTxs
```

where `newTxs` is defined earlier as:

```hs
             let newTxs = fold $ fst <$> cps 
```

The meta only updates the progress, so we're fine updating only the last one for the last block applied (as we did before). Anyway, good catch, but I am slightly worried that our unit tests didn't quite catch that :no_mouth: ...

KtorZ replied:
Fixed.

jonathanknowles commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -681,7 +683,9 @@ newWalletLayer tracer bp db nw tl = do
         DB.withLock db $ do
-            (cp, meta) <- _readWallet wid
+            (wallet, meta) <- _readWallet wid
+            let cps = applyBlocks @s @t blocks wallet
```
I would like to suggest using `txs_cps` here, as it conveys the fact that we have a list of **pairs** of transactions (`txs`) and checkpoints (`cps`), rather than just a list of checkpoints.

KtorZ replied:
This `_` really troubled me to be honest :sweat_smile: ... Plus, txs are somewhat part of the "checkpoint" concept. I do agree however that we could better convey the underlying data structure with a more adequate name. `txsCp` will do I suppose?

jonathanknowles replied:
The issue I have with something like `txsCp` is that it seems to imply a singular item, when in fact we a list of things.

This goes against the common convention for using the `s` suffix to denote a plural. So for example, `ks` for a list of keys, and `vs` for a list of values, `kvs` for a list of key-value pairs, and so on.

If we have to use camelCase, then perhaps `txsCps`?

KtorZ replied:
I went for: 

```hs
            let (txs, cps) = NE.unzip $ applyBlocks @s @t blocks wallet
```

Since in the end, most subsequent calls where doing stuff like `fst <$> ...` or `snd <$> ...`

jonathanknowles reviewed (dismissed):
I agree it's a good idea to avoid writing checkpoints for stable blocks.

However, I'm not sure that this implementation is correct: we still need to write transaction data to the DB for stable blocks, but this change appears to prevent this from happening, unless I've misunderstood?

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -692,34 +696,40 @@ newWalletLayer tracer bp db nw tl = do
+            -- We define unstable blocks as blocks that `k` blocks from the
+            -- network tip, where k is the epoch stability parameter.
+            let (Quantity k) = epochStability
+            let bhUnstable = fromIntegral (flatSlot epochLength nodeTip) - k
```
🤔 This appears to be mixing up slots and blocks. I think we should compare against the blockHeight of the tip minus k. 

(See description of #640)

KtorZ replied:
Good point!

KtorZ replied:
Fixed!

KtorZ committed:
re-generate nix machinery

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -692,34 +695,45 @@ newWalletLayer tracer bp db nw tl = do
+            -- We cast `k` and block heights to 'Integer' since at a given point
+            -- in time, `k` may be greater than the tip.
+            let (Quantity k) = epochStability
+            let (Quantity tipBh) = blockHeight cpLast
```
I wonder if it might not turn out to be a bit tricky. `cpLast` shouldn't necessarily be the network tip, because `restoreBlocks` is called with batches of blocks.

Restore benchmarks seem slower, I believe because of above:

```
All results:
  restore testnet seq: 40.67 s
  restore testnet 10% ownership: 40.86 s
Benchmark restore: FINISH
Completed 2 action(s).
```

But I'm not on the clear with how all the pieces should fit together in the grand scheme of things / if @rvl had any specific plans for this

KtorZ replied:
Man.. Sorry, I am distracted this morning. So, yes, of course the last checkpoint in a particular batch isn't the network tip >.> ... Now, yeah, this is why I started using the `nodeTip` initially but then, we only have its slot number and not its block number. 

I am not sure what is safe to assume here. In practice we won't have much empty blocks but they can happen. In theory, it can even be that 2160 blocks are empty which means that rolling back of `k` blocks would require a checkpoint very far away. 

Yet, I believe we can sort it out nicely by: 

- Making regular sparse checkpoints (as this does), every batch
- Having indeed the last `k` checkpoints for each slots near the network tip
- Upon rollback, selecting the nearest checkpoint. 

In practice, we should always end up having enough checkpoints, even if there's like only 10% of blocks because, big rollbacks are less likely than small ones. If it happens, then we'll simply rollback to the nearest checkpoint and recover from there, even if it's many blocks from the hypothetical point of intersection.

KtorZ replied:
We could also lookup the node's own chain height: 
 
https://input-output-hk.github.io/jormungandr/jcli/rest.html#node-stats

Anviking replied:
> We could also lookup the node's own chain height:

Hm… That's interesting

For me it seems very empty though:
```
blockRecvCnt: 0
lastBlockFees: 0
lastBlockSum: 0
lastBlockTime: ~
lastBlockTx: 0
txRecvCnt: 0
uptime: 207
```
despite the tip blockId changing.

I also imagined we could check if the last blockId in the batch matches the blockId of the network tip. But then we might discover too late that the unstable blocks started in the previous batch…

KtorZ replied:
From the rust team: 

- Chain stats is currently not well updated 
- The `blockRecvCnt` does not represent the chain height, but instead, the total number of blocks received from all peers (so if 10 peers sends 10 blocks each, `blockRecvCnt` equals 100).

KtorZ replied:
Also, the binary block header format actually contains the chain length:

```
Block Header

The header is a small piece of data, containing enough informations for validation and network deduplication and a strong signed cryptographic link to the content.

Common (2 * 64 bits + 1 * 32 bits + 2 * 256 bits = 84 bytes):
    Size of Header: 2 bytes (16 bits): Maximum header is thus 64K not including the block content
    Version of block: 2 bytes (16 bits)
    Size of Content: 4 bytes (32 bits)
    Block Date: Epoch (4 bytes, 32 bits) + Slot-id (4 bytes - 32 bits)
    Chain length (number of ancestor blocks; first block has chain length 0): 4 bytes (32 bits)
    Hash of content H(Content) (32 bytes - 256 bits)
    Parent Header hash : 32 bytes (256 bits)
```

Anviking replied:
😮 !

KtorZ replied:
so, I've extracted the `chainLength` from the tip block header in Jormungandr network layer, such that we now have:

```hs
    , networkTip
        :: ExceptT ErrNetworkTip m (BlockHeader, Quantity "block" Natural)
```

instead of just:

```hs
    , networkTip
        :: ExceptT ErrNetworkTip m BlockHeader
```

For `http-bridge`, it "defaults" to the node's flat slot which should be good-enough. Now, we can reliably use the node's chain height during restoration to figure out whether we are in an unstable zone.

KtorZ committed:
return (and use) node block height alongside the network tip

Anviking reviewed (approved):
lgtm!

KtorZ committed:
refactor BlockchainParameters declaration in WalletSpec

rvl commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -446,10 +445,12 @@ data BlockchainParameters t = BlockchainParameters
         -- ^ Number of slots in a single epoch.
     , getTxMaxSize :: Quantity "byte" Word16
         -- ^ Maximum size of a transaction (soft or hard limit).
+    , getEpochStability :: Quantity "block" Word32
```
"Epoch stability" seems a bit imprecise. It's the ouroboros security parameter _k_ here, isn't it? That has nothing to do with epochs, except that they currently happen to be _10×k_ slots in length. Could we call it "max unstable blocks" (the maximum number of blocks that could potentially be rolled back, a.k.a. "k") instead?

KtorZ replied:
The term ia borrowed from Jormungandr's terminology. Happy to give a more significant one though. I wonder however if "max unstable block" isn't too precise.

rvl commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -692,34 +701,41 @@ newWalletLayer tracer bp db nw tl = do
+            let (Quantity k) = epochStability
+            let bhUnstable :: Integer
+                bhUnstable = fromIntegral nodeHeight - fromIntegral k
+            forM_ (NE.init cps) $ \cp -> do
```
Would it be possible to define a separate pure function that filters `cps` into a batch? Then this can more easily be tested. Something like:
```haskell
let cps' = batchCheckpoints k networkHeight cps
mapM (DB.putCheckpoint db (PrimaryKey wid)) cps'
```

KtorZ replied:
Makes sense. Maybe in the form of an iterator "forEachUnstable action"..

KtorZ committed:
Store checkpoints in order (last stored last)

At the moment, the db layer removes previous checkpoints when storing
a new one. So, in the event where we would store many checkpoints, the
one we want to survive must be stored last.
